### DISCUSSION TITLE
URL Redirection Port Handling Differences Between Next.js v12 and v13

### DISCUSSION THREAD

**Discussion by BO-LIKE-CHICKEN** [vercel/next.js#68089] [Help] [answered]
## Description:
When setting the port to '80' for URL redirection, Next.js v12 does not append the port to the URL, while Next.js v13 does append it.

## To Reproduce:

1. Create a new Next.js project with version 12.2.6.
2. Add the following middleware.ts file:
```ts
// middleware.ts
import { NextRequest, NextResponse } from 'next/server';

export function middleware(req: NextRequest) {
  const url = req.nextUrl.clone();
  url.port = '80';

  const hostUrl = 'https://foo.com';

  return NextResponse.redirect(url);
   // Redirect to '' https://foo.com" in v12
   // Redirect to '' https://foo.com:80" in v13
}
```
3. Observe the URL redirection in the browser or logs. Notice that the URL does not include the port '80'.
4. Repeat the same steps in a Next.js project with version 13.5.6.
5. Observe that the URL now includes the port '80' in the redirection.

## Expected Behavior:
The behavior should be consistent across versions. 

## Environment:
Next.js Version: 12.2.6 and 13.5.6

**Comment by icyJoseph** [comment, marked as answer]
I wonder if this is related to a quirk of the URL constructor.

```js
let url = "http://example.com:80"
new URL(url).port // ""
```

Same for https and 443:

```js
let url = "https://example.com:443"
new URL(url).port // ""
```

Try other ports with those protocols, and it'd be fine, but that combination of http -> 80 and https -> 443, has this behavior.

> If an input string passed to the [URL()](https://developer.mozilla.org/en-US/docs/Web/API/URL/URL) constructor doesn't contain an explicit port number (e.g., https://localhost) or contains a port number that's the default port number corresponding to the protocol part of the input string (e.g., https://localhost:443), then in the [URL](https://developer.mozilla.org/en-US/docs/Web/API/URL) object the constructor returns, the value of the port property will be the empty string: ''.

- https://developer.mozilla.org/en-US/docs/Web/API/URL/port

As per spec, https://url.spec.whatwg.org/#url-miscellaneous:

Special scheme | Default port
-- | --
"ftp" | 21
"file" | null
"http" | 80
"https" | 443
"ws" | 80
"wss" | 443

So `new URL("ftp://example.com:21").port` gives empty string too. Same for, `new URL("ws://example.com:80").port`, and `new URL("wss://example.com:443").port`.

I'd guess that between v12 and v13, the URL constructor was implemented into those functions. Try with other ports, and the same protocol, if the port starts to appear, then the URL constructor is at the heart of the change in behavior.

**Reply by BO-LIKE-CHICKEN** [reply]
Thank you for the detailed response! Your explanation about the URL constructor and the default ports for specific protocols helped me understand the root cause of the issue.
